File: Python/tagiofuns/sens2var.py
def sens2var(Sx=None, Sy=None, r=None):
    
    """Extract data from a sensor dictionary.
    
    :param Sx: dictionary you want data from
    :type: must be a dictionary
    :param Sy: dictionary you want data from
    :type: must be a dictionary
    :param r: can stand in for fs or T:
        fs: sampling rate, a scalar
        T: sample times, a vector
    
    Ways to call the function:
       X,fs = sens2var(Sx)     % regularly sampled data
       or
       X,T = sens2var(Sx)      % irregularly sampled data
       
    Can also be called with two variables, in which case, they are checked for compatibility (i.e., same length and sampling):
       X,Y,fs = sens2var(Sx,Sy); % regularly sampled data
       or
       X,Y,T = sens2var(Sx,Sy);  % irregularly sampled data
       
    Can also be called with a trailing string 'regular' to check if the sensor dictionaries are regularly sampled. If not, X will be
    returned empty.
    
    :raises TypeError: 
    Sx, Sy must be sensor dictionaries. If not X will be returned empty.
    
    :returns:
        A tuple where:
            X, Y are arrays of sensor data. These are in the unit and frame as stored in the input sensor dictionaries.
            fs is the sampling rate of the sensor data in Hz (samples per second).
            T is the time in seconds of each measurement in data for irregularly sampled data. The time reference (i.e., the 0 time) is with
            respect to the start time of the data in the sensor dictionary.
    
    
    Example:
    load_nc('testset3')
    [pca_dur,pca_time] = sens2var(PCA)
    
    Valid: Python
    markjohnson@st-andrews.ac.uk; dmwisniewska@gmail.com
    Last modified: 08 July 2021

    """

    import numpy as np
    
    X, Y, fs = ([] for i in range(3))

    if not Sx:
        print(help(sens2var))
        return (X,Y,fs)
    
    if not (isinstance(Sx,dict) and ('data' in Sx.keys()) and ('sampling' in Sx.keys())):
        print(' Error: input argument must be a sensor dictionary\n')
        return (X,Y,fs)
    
    if Sy:
        if isinstance(Sy,str):
            r = Sy
            Sy = None
    
    if Sy:
        if not isinstance(Sy,dict) or 'data' not in Sy.keys() or 'sampling' not in Sy.keys():
            print(' Error: input argument must be a sensor dictionary\n')
            return (X,Y,fs)
    
    R = [Sx['sampling']=='regular']*2
    if Sy:
        R[1] = Sy['sampling']=='regular'
    if r and r=='regular' and sum(R)<2:
        print(' Error: input argument must be regularly sampled\n')
        if Sy:
            return (X,Y,fs)
        else:
            return (X,fs)
        return
    
    if sum(R)==1:
        print(' Error: input arguments must both be sampled in the same way\n')
        return (X,Y,fs)
    
    if R[0]:
        X = Sx['data']
        fs = Sx['sampling_rate']
    else:
        fs = Sx['data'][:,0]
        if Sx['data'].size/Sx['data'].shape[0]>1:
            X = Sx['data'][:,1:]
        else:
            X = np.ones((len(fs),1))

    if not Sy:
        return (X,fs)
    
    # here on for two input variables
    if R[1]:
        if fs != Sy['sampling_rate']:
            print(' Error: input arguments must both have the same sampling rate\n')
            X, Y, fs = ([] for i in range(3))
            return (X,Y,fs)
        Y = Sy['data']

    else:
        if Sy['data'].size/Sy['data'].shape[0]>1:
            Y = Sy['data'][:,1:]
        else:
            Y = np.ones((len(Sy['data']),1))

    if X.shape[0] != Y.shape[0]:
        print(' Error: input arguments must both have the same number of samples\n')
        X, Y, fs = ([] for i in range(3))
        return (X,Y,fs)  
    
    return (X,Y,fs)

File: matlab/tagiofuns/sens2var.m
function    [X,Y,fs] = sens2var(Sx,Sy,r)

%    [X,fs] = sens2var(Sx)     % regularly sampled data
%    or
%    [X,T] = sens2var(Sx)      % irregularly sampled data
%    Extract data from a sensor structure.
%    Can also be called with two variables, in which case, they
%    are checked for compatibility (i.e., same length and sampling):
%    [X,Y,fs] = sens2var(Sx,Sy); % regularly sampled data
%    or
%    [X,Y,T] = sens2var(Sx,Sy);  % irregularly sampled data
%    Can also be called with a trailing string 'regular' to check if
%    the sensor structures are regularly sampled. If not, X will be
%    returned empty.
%
%    Inputs:
%    Sx, Sy must be sensor structures. If not X will be returned empty.
%
%    Returns:
%    X, Y are vectors or matriced of sensor data. These are in the unit and
%     frame as stored in the input sensor structures.
%    fs is the sampling rate of the sensor data in Hz (samples per second).
%    T is the time in seconds of each measurement in data for irregularly
%     sampled data. The time reference (i.e., the 0 time) is with
%     respect to the start time of the data in the sensor structure.
%
%    Example:
%     load_nc('testset3')
%     [pca_dur,pca_time] = sens2var(PCA)
%
%    Valid: Matlab, Octave
%    markjohnson@st-andrews.ac.uk
%    Last modified: 22 July 2017

X = [] ; Y = [] ; fs = [] ;
if nargin<1,
   help sens2var
   return
end

if ~(isstruct(Sx) && isfield(Sx,'data') && isfield(Sx,'sampling')),
   fprintf(' Error: input argument must be a sensor structure\n') ;
   return
end

if nargin==1,
   r = [] ; Sy = [] ;
elseif nargin==2
   if ischar(Sy),
      r = Sy ;
      Sy = [] ;
   else
      r = [] ;
   end
end

if nargin>1 && ~isempty(Sy),
   if ~isstruct(Sy) || ~isfield(Sy,'data') ~isfield(Sy,'sampling')
      fprintf(' Error: input argument must be a sensor structure\n') ;
      return
   end
end

R = [1 1]*strcmp(Sx.sampling,'regular') ;
if ~isempty(Sy),
   R(2) = strcmp(Sy.sampling,'regular') ;
end
if ~isempty(r) && strcmpi(r,'regular') && sum(R)<2,
   fprintf(' Error: input argument must be regularly sampled\n') ;
   return
end 

if sum(R)==1,
   fprintf(' Error: input arguments must both be sampled in the same way\n') ;
   return
end 
   
if R(1),
   X = Sx.data ;
   fs = Sx.sampling_rate ;
else   
   fs = Sx.data(:,1) ;
   if size(Sx.data,2)>1,
      X = Sx.data(:,2:end) ;
   else
      X = ones(length(fs),1) ;
   end
end

if isempty(Sy),
   Y = fs ;
   return
end
   
% here on for two input variables
if R(2),
   if fs ~= Sy.sampling_rate,
      fprintf(' Error: input arguments must both have the same sampling rate\n') ;
      X = [] ; Y = [] ; fs = [] ;
      return
   end
   Y = Sy.data ;
      
else
   if size(Sy.data,2)>1,
      Y = Sy.data(:,2:end) ;
   else
      Y = ones(size(Sy.data,1),1) ;
   end
end

if size(X,1)~=size(Y,1),
   fprintf(' Error: input arguments must both have the same number of samples\n') ;
   X = [] ; Y = [] ; fs = [] ;
   return
end
   
